### PR TITLE
Introduce integration mode of sanitycheck

### LINKED ISSUE #25104: whitelist in {sample,testcase}.yaml precludes a test from being run with sanitycheck
Currently the `whitelist` keyword in a sample's or test's yaml file means 2 things:

- This sample or test can only be ever run on this particular platform(s)
- sanitycheck will only build and run this sample or test on this particular platform(s) in CI (PR or nightly CI)

I think we need to decouple the two meanings, perhaps by introducing:

- `compatible`: All platforms this sample or test is compatible with
- `ci-whitelist`: All platforms that this sample or test will be built and run on during CI

That way a `sanitycheck -p <platform>` would run on all samples and tests that are compatible, but a `sanitycheck -p <platform> --ci` would use the whitelist instead.

### PR DESCRIPTION
**Pull request by nashif**
his adds a new option, integration which is used to list platforms to use in integration testing (CI) and avoids whitelisting platforms. This will enable running those tests on real hardware if the --integration option is omitted.
--integration option is now not default, in the future this will be the default and will build/test a minimal set, running sanitycheck with more coverage will be possible use another option.

oh, one more thing in this PR: use native_posix exclusively where it makes sense and avoid massive build/running trivial stuff that is being tested elsewhere anyways..

resolves #25104.

### CODE REVIEW

File: tests/benchmarks/mbedtls/testcase.yaml [hunk 1]
```diff
@@ -3,4 +3,6 @@ common:
   tags: crypto
 tests:
   benchmark.crypto.mbedtls:
-    platform_whitelist: qemu_x86 frdm_k64f sam_e70_xplained nrf52840dk_nrf52840
+    filter: CONFIG_TEST_RANDOM_GENERATOR
+    integration:
+      - qemu_x86
```
Comment (stephanosio) [inline tests/benchmarks/mbedtls/testcase.yaml:8]: Is there a reason `frdm_k64f sam_e70_xplained nrf52840dk_nrf52840` are removed from the integration testing list?

I can see the value of running these "integration tests" on them, since `mbedtls` is a crypto library and should ideally be tested/validated on every major supported platform if not all.

File: samples/bluetooth/mesh_demo/prj_nrf51_blenano.conf [hunk 2]
```diff
@@ -18,16 +17,3 @@ CONFIG_BT_MESH_MODEL_GROUP_COUNT=2
...
-#CONFIG_BT_MESH_IV_UPDATE_TEST=y
-#CONFIG_BT_MESH_DEBUG=y
-#CONFIG_BT_MESH_DEBUG_NET=y
-#CONFIG_BT_MESH_DEBUG_TRANS=y
-#CONFIG_BT_MESH_DEBUG_PROV=y
-#CONFIG_BT_MESH_DEBUG_BEACON=y
-#CONFIG_BT_MESH_DEBUG_CRYPTO=y
-#CONFIG_BT_MESH_DEBUG_ADV=y
-#CONFIG_BT_MESH_DEBUG_ACCESS=y
-#CONFIG_BT_DEBUG_LOG=y
```
Comment (stephanosio) [inline samples/bluetooth/mesh_demo/prj_nrf51_blenano.conf:33]: I wonder if these were intentionally left here so as to remind users of the existence of these potentially useful/relevant options (after all, this is a "sample").

Comment (nashif) [reply samples/bluetooth/mesh_demo/prj_nrf51_blenano.conf:33]: those commented options bitrot really fast and become irrelevant if they are not tested. If there is anything to remind the user, it should be part of the documentation. I argue this was just left over from someone debugging the sample.

File: tests/benchmarks/mbedtls/testcase.yaml (hunk 1 above)
Comment (nashif) [reply tests/benchmarks/mbedtls/testcase.yaml:8]: > I can see the value of running these "integration tests" on them, since `mbedtls` is a crypto library and should ideally be tested/validated on every major supported platform if not all.

validated how? We are just building those tests right now, integration testing limits that to only 1 runnable platform in CI. Removing the whitelisting here enables this on every single platform we have including those that were whitelisted and those that were not.

Comment (stephanosio) [reply tests/benchmarks/mbedtls/testcase.yaml:8]: > validated how? We are just building those tests right now, integration testing limits that to only 1 runnable platform in CI. Removing the whitelisting here enables this on every single platform we have including those that were whitelisted and those that were not.

@nashif I was under the impression that the `integration:` platform list was migrated with the future support for hardware CI testing in mind (i.e. I see other commits in this PR that left `frdm_k64f` and `sam_e70_xplained` in the `integration:` list).

File: scripts/sanity_chk/sanitylib.py [hunk 3]
```diff
@@ -1208,12 +1208,6 @@ def get_test(self, name, valid_keys):
             d[k] = v
 
         for k, v in self.tests[name].items():
-            if k not in valid_keys:
-                raise ConfigurationError(
-                    self.filename,
-                    "Unknown config key '%s' in definition for '%s'" %
-                    (k, name))
-
```
Comment (PerMac) [inline scripts/sanity_chk/sanitylib.py:1216]: What is the reason behind this change? Is it because 'integration' is not in the list of valid_keys? Why not adding 'integration' there and keeping this check in place?

Comment (nashif) [reply scripts/sanity_chk/sanitylib.py:1216]: this is already validated by the yaml schema checks, it is just an old check we had that is duplicating what yaml verification does now.

File: tests/benchmarks/mbedtls/testcase.yaml (hunk 1 above)
Comment (PerMac) [reply tests/benchmarks/mbedtls/testcase.yaml:8]: If I understood correctly from the discussion about the `integration` list it supposes to contain only emulated platforms where tests/samples will be run by zephyr's CI. I've seen `fedm_k64f` and `sam_e70_xplained ` added to the integration list in cmsis_dsp tests but (again if I got it correct from @nashif description) it was due to a 'hassle' with being able to add the `integration` list and all the samples/tests will have to be still reevaluated to correctly handle the `integration`.

**Review by PerMac — APPROVED** [review]

**Review by galak — DISMISSED** [review]
Need to update doc/guides/test/sanitycheck.rst.

I wish we had a better name than `integration`, best I can come up with is `ci`.

**Comment by andyross** [comment]
> I wish we had a better name than integration

Interestingly, the most common term of art in the industry for this kind of simplest-possible-but-still-comprehensive test to check for obvious bugs before merge is a "sanity check".  Though I'm reliably told that in the modern world that language is frowned on as not particularly inclusive.

Seriously, I've been staying out of this one for obvious reasons but it all looks good to me.

**Comment by galak** [comment]
> Interestingly, the most common term of art in the industry for this kind of simplest-possible-but-still-comprehensive test to check for obvious bugs before merge is a "sanity check". Though I'm reliably told that in the modern world that language is frowned on as not particularly inclusive.
> 
> Seriously, I've been staying out of this one for obvious reasons but it all looks good to me.

I wasnt thinking of it in those terms, just that integration didn't spring to mind the meaning of build/run this test for CI purposes.

**Comment by stephanosio** [comment]
> I wasnt thinking of it in those terms, just that integration didn't spring to mind the meaning of build/run this test for CI purposes.

We have `platform_whitelist` to list the platforms that are supported by a test/sample, I think something like `ci_whitelist` would be more intuitive.

**Comment by PerMac** [comment]
Some ideas for the name (I tried to avoid uninclusive language).
CI_smoketest, CI_BVT (build verification test), CI_verification, CI_smokelist, CI_scope. The CI prefix is optional, however, I think using it clearly shows the purpose of the list.

**Comment by nashif** [comment]
> I wish we had a better name than `integration`, best I can come up with is `ci`.

what is the i in ci? :)

**Comment by galak** [comment]
> what is the i in ci? :)

I know :)

File: tests/benchmarks/mbedtls/testcase.yaml (hunk 1 above)
Comment (ceolin) [reply tests/benchmarks/mbedtls/testcase.yaml:8]: That is not clear for me. What is the criteria used to pick where to do the integration test ?

Related to this specific test,it should be tested in different platforms, mbedtls has some integration with platform that should be tested as well, for example random number generation. This benchmarking test is exercising several functions and would be nice to have them running in as much targets as possible.

**Comment by ceolin** [comment]
> > I wish we had a better name than `integration`, best I can come up with is `ci`.
> 
> what is the i in ci? :)

I have no problem with the name, the only thing I can think against is that this may bring some confusion with unit and integration tests, that in this cases refers to tests different components integration.

**Comment by ceolin** [comment]
This option is not a replacement for `platform_whitelist` right ? In this case why `platform_whitelist` as removed in some commits ?

**Comment by PerMac** [comment]
> This option is not a replacement for `platform_whitelist` right ? In this case why `platform_whitelist` as removed in some commits ?

I believe the reason for this PR is the fact that whitelisting became an issue blocking proper on-target testing. If a whitelist exists it prevents the test from being run on any other platform, which is especially a problem in the context of test on the physical boards. Whitelisting ended up as a simple way to provide minimum test coverage (e.g. one writes a test based on a platform on his/her desk and adds the platform to the whitelist, which prevents the test from being run on any other platform) and also a list to minimize the scope of the CI testing to a limited subset of platforms. I believe the trend is to depreciate whitelisting and decouple 2 issues: the scope for the CI tests is given by the 'integration' list and the filtering of the tests runnability on a given platform will be based on dts compatibility.

**Comment by carlescufi** [comment]
Also not a fan of `integration` but cannot come up with something better either. Perhaps `ci` would be shorter and easier to understand?

Also, this closes #25104 right @nashif?

**Comment by nashif** [comment]
> Also not a fan of `integration` but cannot come up with something better either. Perhaps `ci` would be shorter and easier to understand?

Again, ci is 'continuous integration', this is 'integration', so not sure what the issue is here.. This can also be run out of CI if someone wants a faster run locally, so it is not limited to ci. 

and yes, it fixes  #25104.

File: doc/guides/test/sanitycheck.rst [hunk 4]
```diff
@@ -303,7 +303,16 @@ arch_exclude: <list of arches, such as x86, arm, arc>
     Set of architectures that this test case should not run on.
 
 platform_whitelist: <list of platforms>
-    Set of platforms that this test case should only be run for.
+    Set of platforms that this test case should only be run for. Do not use
+    this option to limit testing or building in CI due to time or resource
+    constraints, this option should only be used of the test or sample can
```
Comment (galak) [inline doc/guides/test/sanitycheck.rst:308]: `used of` I think is suppose to be `used if`.

File: doc/guides/test/sanitycheck.rst [hunk 5]
```diff
@@ -303,7 +303,16 @@ arch_exclude: <list of arches, such as x86, arm, arc>
...
+    only be run on the whitelisted platform and nothing else.
+
+integration: <YML list of platforms/boards>
```
Comment (galak) [inline doc/guides/test/sanitycheck.rst:311]: Maybe `platform_integration_list` or `integration_platforms` instead of `integration`?

**Review by galak — DISMISSED** [review]
* Minor typo in docs.
* Idea of using a slightly more descriptive term in YAML.

File: doc/guides/test/sanitycheck.rst (hunk 5 above)
Comment (nashif) [reply doc/guides/test/sanitycheck.rst:311]: Another proposal: drop all of this and do it completely differently...

basically the option would be --mode with a list of modes that are supported, for example:
``--mode [daily|pull-request|default|...]```


now, in the testcase yaml file this can be triggered with:

```

 test:
   mode:
     pull-request:
       - <platform>
       - <platform>
     daily:
       - all
```
Still need to figure out the syntax and allow flexibility here.

Existing `build_on_all` would fold into that and removed as a standalone option for example.

Comment (nashif) [reply doc/guides/test/sanitycheck.rst]: ok, sticking with original proposal with a name change, now using integration_platforms.

File: tests/drivers/dac/dac_loopback/testcase.yaml [hunk 6]
```diff
@@ -1,6 +1,5 @@
 common:
-  tags: dac drivers userspace
+  tags: dac drivers
 tests:
   drivers.dac:
-    platform_whitelist: frdm_k64f nucleo_l073rz nucleo_l152re twr_ke18f
```
Comment (erwango) [inline tests/drivers/dac/dac_loopback/testcase.yaml:5]: This won't be sufficient.
In `tests/drivers/dac/dac_loopback/src/test_dac.c` there is some board specific configuration code, that ends up in:
```
#else
#error "Unsupported board."
#endif
```
So I think that with current implementation, the platform_withelist should stand.

**Review by erwango — CHANGES_REQUESTED** [review]

File: tests/drivers/dac/dac_loopback/testcase.yaml (hunk 6 above)
Comment (nashif) [reply tests/drivers/dac/dac_loopback/testcase.yaml:5]: this depends on dac already, so that should be fine.

Comment (erwango) [reply tests/drivers/dac/dac_loopback/testcase.yaml:5]: So each board declaring dac support will have to go and edit this test_dac.c, but maybe at the end someone will try to find a smarter test implementation... ok

**Review by erwango — APPROVED** [review]

**Review by LeiW000 — APPROVED** [review]
It looks good to me.

**Review by ceolin — APPROVED** [review]
